### ISSUE TITLE
Implement "Element is not clickable" autofix

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#2866]
_From @christian-bromann on March 10, 2018 16:38_

## The problem

Sometimes commands on elements like (click) can't be executed due to the "Element is not clickable at xy" exception. Let's implement a couple of workarounds (explained in [this great SO post](https://stackoverflow.com/a/44916498/765174)) as part of the [element handler error middleware](https://github.com/webdriverio/v5/blob/master/packages/webdriverio/src/middlewares.js#L17) to prevent users from running into this issue.

_Copied from original issue: webdriverio/v5#21_

**Comment by christian-bromann** [comment]
_From @seadb on March 27, 2018 19:0_

I will be sew happy when this is implemented :)

**Comment by christian-bromann** [comment]
👏 nice, let me know if you have any questions on this

**Comment by christian-bromann** [comment]
_From @thephilwells on April 3, 2018 20:33_

Is this issue claimed? I'd like to give it a shot if I may.

**Comment by christian-bromann** [comment]
@seadb any progress on this? Or can @thephilwells take over?

**Comment by christian-bromann** [comment]
Let's ensure we include the work that is already done by @thephilwells

**Comment by christian-bromann** [comment]
Removing it from the v5 release as it is currently also not implemented in v4 and I want to get v5 out asap.

**Comment by thephilwells** [comment]
Makes sense. Apologies for dragging my feet on this.

**Comment by christian-bromann** [comment]
@thephilwells feel free to propose a PR

**Comment by WillBrock** [comment]
@abjerstedt is this already implemented in v5?

**Comment by christian-bromann** [comment]
@WillBrock nope, it should be implemented in the webdriverio middleware. There are a couple of steps we can try to avoid that error, e.g. scroll element into the center of the viewport. However we should not try to remove the element or move it somewhere else because we shouldn't reject the meaning of that error.

**Comment by mgrybyk** [comment]
@CrispusDH is this one fixed by your PR?

**Comment by CrispusDH** [comment]
@mgrybyk it is one piece of SO answer. Also, my workaround will work if overlapping element changes position during scrolling. But in this specific case it works for me. I have implement example (link in PR I think).

**Comment by christian-bromann** [comment]
I think there are various scenarios where an element is not clickable (see issues above).

**Comment by erwinheitzman** [comment]
Just want to get something clear.
Are we allowing elements to be clicked that in reality could never be clicked (due to overlapping elements)?

**Comment by christian-bromann** [comment]
> Are we allowing elements to be clicked that in reality could never be clicked (due to overlapping elements)?

I would say no! But we should make the error message more clear and retry the click based on different conditions (after scrolling and centering it)

**Comment by Shreya76** [comment]
Could you please share the code for that, because I have tried so many ways by scrolling, still it does not work.

clickOn(element) {
    $(element).waitForEnabled(maxWaitTime);
    let coordinates= $(element).getLocation();
    browser.execute('document.elementFromPoint("'+coordinates.x+'","'+coordinates.y+'").click()');
    //browser.scroll(coordinates.x, coordinates.y);  //this scrolls but still the same issue after clicking
    //browser.pause(100);
    //$(element).click();
    //$(element).click({x: 0}, {y: 30});  // this throws error as element.clice is not allowed
  }

**Comment by mgrybyk** [comment]
@Shreya76 can you please try newly added command `isClickable`? You may wait for element to be clickable first before clicking on it

**Comment by Shreya76** [comment]
I tried using waitUntil and then isClickable, even this fails and the error message is-  el.isClickable is not a function

**Comment by mgrybyk** [comment]
@Shreya76 update to the latest version all wdio packages

**Comment by Shreya76** [comment]
> @Shreya76 update to the latest version all wdio packages

Okay, I will try that

**Comment by JH1ller** [comment]
@mgrybyk I get 'el.isClickable is not a function' too and I installed all related packages with package@latest.
Here are my versions:

```
    "appium-doctor": "^1.4.3",
    "wdio-appium-service": "^0.2.3",
    "wdio-jasmine-framework": "^0.3.2",
    "wdio-spec-reporter": "^0.1.5",
    "webdriverio": "^5.16.7",
    "appium": "^1.15.1"
```

any ideas?

**Comment by mgrybyk** [comment]
@JH1ller use gitter for such questions. You have invalid project setup, consider using getting started guide

**Comment by chirag-farfetch** [comment]
any luck on this, i am also starting to have issues with this?

**Comment by mgrybyk** [comment]
Use wait for clickable command

**Comment by sethuster** [comment]
The wait for clickable command seems solve the original problem.  Do we need this issue open any longer?

**Comment by christian-bromann** [comment]
I think it would be nice to have an autofix for it. You might forget calling the command and some positions and having WebdriverIO actively providing a safety net for it would add value for users.

**Comment by sethuster** [comment]
I'm going to give this a shot...

**Comment by christian-bromann** [comment]
@sethuster any updates? If there is anything I can do to help, please let me know!
